### PR TITLE
Implement TWTS-laser twtstight to replace twtsfast

### PR DESCRIPTION
**Pull request by BeyondEspresso**
This PR implements the new laser model for TWTS, called twtstight. It replaces twtsfast. The new implementation is a completely new model as it is fully Maxwell-compatible, while also supporting narrower (i.e. tighter) focal line heights down to the lambda-scale. Additionally, different linear polarizations are natively parameterized, such that the number of laser-kernels to call is oftenreduced by a factor of 2. The PR compiles and has been excessively tested (successfully) against a Mathematica implementation. However, the unit-tests have been removed here, since these need to be implemented in slightly recasted form at another place in PIConGPU. This will be the goal of a future PR. The FOM-benchmark and Bunch example have been updated, since twtstight replaces twtsfast.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
please rebase and fix the CI issue

```
/root/buildCI/params/TWEAC-FOM/cmakePreset_0/include/picongpu/param/TwtsBackgroundLaser.param:27:10: fatal error: 'picongpu/param/grid.param' file not found
#include "picongpu/param/grid.param"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by BeyondEspresso** [comment]
@psychocoderHPC Done.

**Comment by psychocoderHPC** [comment]
@PrometheusPi @steindev one of you should review this PR

File: include/picongpu/fields/background/templates/twtstight/BField.hpp [hunk 1]
```diff
@@ -9,7 +9,7 @@
  *
  * PIConGPU is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/BField.hpp]: Superfluous space.

File: include/picongpu/fields/background/templates/twtstight/BField.hpp [hunk 2]
```diff
@@ -104,14 +89,16 @@ namespace picongpu
                  * @param w_x beam waist: distance from the axis where the pulse electric field
                  *  decreases to its 1/e^2-th part at the focus position of the laser [m]
                  * @param phi interaction angle between TWTS laser propagation vector and
-                 *  the y-axis [rad, default = 90.*(PI/180.)]
+                 *  the y-axis [rad, default = 90. * (PI/180.) ]
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/BField.hpp]: Superfluous space before ']'.

File: include/picongpu/fields/background/templates/twtstight/BField.hpp [hunk 3]
```diff
@@ -104,14 +89,16 @@ namespace picongpu
...
                  * @param beta_0 propagation speed of overlap normalized to
                  *  the speed of light [c, default = 1.0]
                  * @param tdelay_user manual time delay if auto_tdelay is false
                  * @param auto_tdelay calculate the time delay such that the TWTS pulse is not
                  *  inside the simulation volume at simulation start timestep = 0 [default = true]
-                 * @param pol determines the TWTS laser polarization, which is either normal or parallel
-                 *  to the laser pulse front tilt plane [ default= LINEAR_X , LINEAR_YZ ]
+                 * @param polAngle determines the TWTS laser polarization angle with respect to x-axis around
+                 * propagation direction [rad, default = 0. * (PI/180.) ] Normal to laser pulse front tilt plane:
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/BField.hpp]: Superfluous space before ']'.

File: include/picongpu/fields/background/templates/twtstight/BField.hpp [hunk 4]
```diff
@@ -177,13 +164,14 @@ namespace picongpu
                  *  for calculating the field */
                 HDINLINE float_T calcTWTSBy(float3_64 const& pos, float_64 const time) const;
 
+                /* UPDATE NAME  */
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/BField.hpp]: Is this a(n old) todo?

File: include/picongpu/fields/background/templates/twtstight/BField.tpp [hunk 5]
```diff
@@ -0,0 +1,687 @@
...
+                        {
+                            return static_cast<float_X>(calcTWTSBy(pos, time_SI));
+                        }
+                        if constexpr(T_component == 2)
+                        {
+                            return static_cast<float_X>(calcTWTSBz(pos, time_SI));
+                        }
+                    }
+                    // we should never be here
+                    return 0.0_X;
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/BField.tpp]: If you should never arrive at this point, then you could also return NaN or something that clearly indicates that something went wrong instead of a valid value '0'.

File: include/picongpu/fields/background/templates/twtstight/EField.hpp [hunk 6]
```diff
@@ -9,7 +9,7 @@
  *
  * PIConGPU is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/EField.hpp]: Superfluous space.

File: include/picongpu/fields/background/templates/twtstight/EField.hpp [hunk 7]
```diff
@@ -103,14 +89,16 @@ namespace picongpu
                  * @param w_x beam waist: distance from the axis where the pulse electric field
                  *  decreases to its 1/e^2-th part at the focus position of the laser [m]
                  * @param phi interaction angle between TWTS laser propagation vector and
-                 *  the y-axis [rad, default = 90.*(PI/180.)]
+                 *  the y-axis [rad, default = 90. * (PI/180.) ]
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/EField.hpp]: Superfluous space before ']'.

File: include/picongpu/fields/background/templates/twtstight/EField.hpp [hunk 8]
```diff
@@ -103,14 +89,16 @@ namespace picongpu
...
                  * @param beta_0 propagation speed of overlap normalized to
                  *  the speed of light [c, default = 1.0]
                  * @param tdelay_user manual time delay if auto_tdelay is false
                  * @param auto_tdelay calculate the time delay such that the TWTS pulse is not
                  *  inside the simulation volume at simulation start timestep = 0 [default = true]
-                 * @param pol dtermines the TWTS laser polarization, which is either normal or parallel
-                 *  to the laser pulse front tilt plane [ default= LINEAR_X , LINEAR_YZ ]
+                 * @param polAngle determines the TWTS laser polarization angle with respect to x-axis around
+                 * propagation direction [rad, default = 0. * (PI/180.) ] Normal to laser pulse front tilt plane:
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/EField.hpp]: Superfluous space before ']'.

File: include/picongpu/fields/background/templates/twtstight/EField.tpp [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+                    {
+                        pos[k][i + 1] = eFieldPositions_SI[k][i];
+                    }
+                }
+
+                /* E-field normalized to the peak amplitude. */
+                return float3_X(calcTWTSEx(pos[0], time), calcTWTSEy(pos[1], time), calcTWTSEz(pos[2], time));
+            }
+
+            HDINLINE float3_X EField::operator()(DataSpace<simDim> const& cellIdx, uint32_t const currentStep) const
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/EField.tpp:127]: 1. Can you please add a comment explaining why there are two different implementations of `operator()` for different argument types?
2. Is it correct that the types of `currentStep` differ between the implementations?

File: include/picongpu/fields/background/templates/twtstight/EField.tpp [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+                            return static_cast<float_X>(calcTWTSEy(pos, time_SI));
+                        }
+                        if constexpr(T_component == 2)
+                        {
+                            return static_cast<float_X>(calcTWTSEz(pos, time_SI));
+                        }
+                    }
+
+                    // we should never be here
+                    return 0.0_X;
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/EField.tpp]: Same as in `BField.tpp`. Why not NaN?

File: include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp [hunk 11]
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2014-2023 Alexander Debus
+/* Copyright 2014-2022 Alexander Debus
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp]: As you performed changes in this file,I think you should change the year to 2024, not 2022.

File: include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp [hunk 12]
```diff
@@ -80,18 +80,18 @@ namespace picongpu
                         /* halfSimSize[2] --> Half-depth of simulation volume (in z); By geometric
                          * projection we calculate the y-distance walkoff of the TWTS-pulse.
                          * The abs( )-function is for correct offset for -phi<-90Deg and +phi>+90Deg. */
-                        float_64 const y1 = float_64(halfSimSize[2] * picongpu::sim.si.getCellSize().z())
-                            * math::abs(math::cos(eta));
+                        float_64 const y1
+                            = float_64(halfSimSize[2] * picongpu::SI::CELL_DEPTH_SI) * math::abs(math::cos(eta));
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp]: 1. Why did you change back to `SI::CELL_DEPTH` here.
2. I thought the variable `eta` is not used anymore in the new version of the code?

Is this an old file that accidentally replaced a newer one?

File: include/picongpu/fields/background/templates/twtstight/getFieldPositions_SI.tpp [hunk 13]
```diff
@@ -9,7 +9,7 @@
  *
  * PIConGPU is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/getFieldPositions_SI.tpp]: Space again.

File: include/picongpu/fields/background/templates/twtstight/getFieldPositions_SI.tpp
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/getFieldPositions_SI.tpp:1]: The year should be adjusted.

File: include/picongpu/fields/background/templates/twtstight/numComponents.hpp
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/numComponents.hpp:1]: Please adjust year and white space here, too.

File: include/picongpu/fields/background/templates/twtstight/twtstight.hpp
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/twtstight.hpp:1]: same again

File: include/picongpu/fields/background/templates/twtstight/twtstight.hpp [hunk 14]
```diff
@@ -55,5 +49,16 @@
...
+#include "picongpu/fields/background/templates/twtstight/BField.hpp"
+#include "picongpu/fields/background/templates/twtstight/EField.hpp"
+
+namespace picongpu
+{
+    namespace templates
+    {
+        namespace twtstight
+        {
+            constexpr uint32_t numSigmas = 10;
```
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/twtstight.hpp:74]: Please document this variable.

File: include/picongpu/fields/background/templates/twtstight/twtstight.tpp
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/twtstight.tpp:1]: Year and white space...

File: include/picongpu/unitless/fieldBackground.unitless
Comment (steindev) [inline include/picongpu/unitless/fieldBackground.unitless:1]: Year and add yourself as author, please.

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/fieldBackground.param [hunk 15]
```diff
@@ -44,139 +48,52 @@ namespace picongpu
...
+            , twtsFieldENeg(
+                  Params::FOCUS_Y_SI,
+                  Params::WAVE_LENGTH_SI,
+                  Params::PULSE_LENGTH_SI,
+                  Params::W0_SI,
+                  -Params::PHI,
+                  Params::BETA_0,
+                  Params::TDELAY,
+                  Params::AUTO_TDELAY,
+                  Params::Polarization)
```
Comment (steindev) [inline share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/fieldBackground.param]: The sign of the polarization angle does not need to be flipped together with the sign of the propagation angle `phi`? (Same for B)

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/fieldBackground.param [hunk 16]
```diff
@@ -190,137 +107,53 @@ namespace picongpu
...
-
-            /** unit: none */
-            constexpr float_64 _A0 = 3.25 * 0.01; // reduced for FOM benchmark
-
             /** unit: Volt /meter */
-            const float3_64 invUnitField = float3_64(1.0 / m_unitField[0], 1.0 / m_unitField[1], 1.0 / m_unitField[2]);
+            float3_64 const invUnitField = float3_64(1.0 / m_unitField[0], 1.0 / m_unitField[1], 1.0 / m_unitField[2]);
 
-            /* laser amplitude in picongpu units [ unit: (Volt/meter) / unitField-factor ]
+            /* laser amplitude in picongpu units [ unit: (Volt /meter) / unitField-factor ]
```
Comment (steindev) [inline share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/fieldBackground.param]: Superfluous white space.

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param [hunk 17]
```diff
@@ -67,523 +69,180 @@
...
+                    , twtsFieldEneg(
+                          Params::FOCUS_Y_SI,
+                          Params::WAVE_LENGTH_SI,
+                          Params::PULSE_LENGTH_SI,
+                          Params::W0_SI,
+                          -Params::PHI,
+                          Params::BETA_0,
+                          Params::TDELAY,
+                          Params::AUTO_TDELAY,
+                          Params::Polarization)
```
Comment (steindev) [inline share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param]: Same question on polarization angle as above.

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param [hunk 18]
```diff
@@ -67,523 +69,180 @@
...
-                        * twtsFieldE2.getComponent<T_component>(totalCellIdx, m_currentStep);
+                    float_64 const invUnitField = 1.0 / m_unitField[T_component];
+                    float_X const amplitude = precisionCast<float_X>(Params::AMPLITUDE_SI * invUnitField);
+                    return amplitude * twtsFieldEneg.getComponent<T_component>(totalCellIdx, m_currentStep);
                 }
 #endif
             };
 
-            class FunctorE3
+            class FunctorBpos
```
Comment (steindev) [inline share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param:164]: We could also use the automatic B-Field computation via SVEA in the incidentField method. Or would this not be precise enough?

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param [hunk 19]
```diff
@@ -67,523 +69,180 @@
...
+                    , twtsFieldBneg(
+                          Params::FOCUS_Y_SI,
+                          Params::WAVE_LENGTH_SI,
+                          Params::PULSE_LENGTH_SI,
+                          Params::W0_SI,
+                          -Params::PHI,
+                          Params::BETA_0,
+                          Params::TDELAY,
+                          Params::AUTO_TDELAY,
+                          Params::Polarization)
```
Comment (steindev) [inline share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param]: polarization?

**Review by steindev — CHANGES_REQUESTED** [review]
Good work! Only a few questions and few repeating typos.

File: include/picongpu/fields/background/templates/twtstight/EField.tpp (hunk 9 above)
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/EField.tpp:127]: These are the two interfaces for fieldBackground and incidentField respectively.
1. Done.
2. Yes, see above.

File: include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp (hunk 12 above)
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp]: 1. This happened during a rebase-session. This is now fixed again.
2. The variable "eta" is only defined and used in this method to guess an initial time delay parameter in case it is not already provided by the user.

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param (hunk 18 above)
Comment (BeyondEspresso) [reply share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param:164]: Sure, but this still needs to be tested for precision and compared to the variant with both E- and B-field.

**Comment by BeyondEspresso** [comment]
@steindev I applied your review comments with documentation and also included the math refactoring. The refactored version was tested to give the same result as the prior version.

File: include/picongpu/fields/background/templates/twtstight/BField.hpp (hunk 4 above)
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/BField.hpp]: Yes.

File: include/picongpu/fields/background/templates/twtstight/BField.tpp (hunk 5 above)
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/BField.tpp]: I made it return a NAN.

File: include/picongpu/fields/background/templates/twtstight/EField.tpp (hunk 10 above)
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/EField.tpp]: Done.

File: include/picongpu/fields/background/templates/twtstight/twtstight.hpp (hunk 14 above)
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/twtstight.hpp:74]: Added documentation.

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/fieldBackground.param (hunk 15 above)
Comment (BeyondEspresso) [reply share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/fieldBackground.param]: Good catch.

**Comment by psychocoderHPC** [comment]
I stopped the CI because you must rebase anyway to resolve the merge conflicts

**Comment by BeyondEspresso** [comment]
@psychocoderHPC I'll do the rebase then.

**Comment by BeyondEspresso** [comment]
@steindev @psychocoderHPC Ok, I rebased the PR and squashed the commits. If you want to have separated commits, I could change this again.

**Comment by psychocoderHPC** [comment]
sry @BeyondEspresso please rebase the PR again, there were a few PRs to fix include dependencies.

**Comment by psychocoderHPC** [comment]
if you have the include to `simulation_defines.hpp` in your code note that this file has deleted.
You should include `picongpu/defines.hpp` for definitions of `float_X`, `simDim`.

**Comment by psychocoderHPC** [comment]
I stopped the CI because there are issues in the code

```
In file included from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/background/templates/twtstight/twtstight.tpp:23,
                 from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/unitless/fieldBackground.unitless:25,
                 from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/_defaultUnitless.loader:38,
                 from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/simulation_defines.hpp:50,
                 from /builds/hzdr/crp/picongpu/include/picongpu/fields/EMFieldBase.x.cpp:23:
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/background/templates/twtstight/BField.tpp: In member function 'picongpu::templates::twtstight::BField::float_T picongpu::templates::twtstight::BField::calcTWTSBy(const float3_64&, picongpu::float_64) const':
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/background/templates/twtstight/BField.tpp:290:33: error: variable 'rhom2' set but not used [-Werror=unused-but-set-variable]
  290 |                 complex_T const rhom2 = rhom * rhom;
      |                                 ^~~~~
In file included from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/background/templates/twtstight/twtstight.tpp:24,
                 from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/unitless/fieldBackground.unitless:25,
                 from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/_defaultUnitless.loader:38,
                 from /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/simulation_defines.hpp:50,
                 from /builds/hzdr/crp/picongpu/include/picongpu/fields/EMFieldBase.x.cpp:23:
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/background/templates/twtstight/EField.tpp: In member function 'picongpu::templates::twtstight::EField::float_T picongpu::templates::twtstight::EField::calcTWTSEy(const float3_64&, picongpu::float_64) const':
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/fields/background/templates/twtstight/EField.tpp:429:33: error: variable 'rhom2' set but not used [-Werror=unused-but-set-variable]
  429 |                 complex_T const rhom2 = rhom * rhom;
```

**Comment by BeyondEspresso** [comment]
@psychocoderHPC Ok, fixed.

**Review by psychocoderHPC — APPROVED** [review]

**Comment by BeyondEspresso** [comment]
@steindev @psychocoderHPC @PrometheusPi Is this PR now done or do you need more changes?

File: include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp
Comment (steindev) [inline include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp:1]: The diff still looks like as if a new file has been accidentally replaced by an old file. As this is working code at the moment, we will refactor this in an upcoming PR. No need to do anything now.

**Review by steindev — APPROVED** [review]
All my questions have been resolved. Got to go from my side!

File: include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp
Comment (BeyondEspresso) [reply include/picongpu/fields/background/templates/twtstight/GetInitialTimeDelay_SI.tpp:1]: Hmpfff! I am certain that I have applied the appropriate changes here, but my hunch is that I accidentally edited this file on a different cluster... 😭 Too many terminals open on too few monitors. 😜
